### PR TITLE
[Feature] Dropdown label and placeholder

### PR DESCRIPTION
**Pull request by aappoalander**
## Description

This PR adds mandatory label and related options for dropdown. In addition, name prop is renamed to visiblePlaceholder and is now ignored by screen readers.

## Motivation and Context

The dropdown component was not accessible due to lack of label. Label is now added as mandatory prop. Label can be hidden, when it's only available for screen readers. When label is hidden, a visualPlaceholder or other means should be used to communicate the purpose of the dropdown for non screen reader users.

Name prop is now renamed as visualPlaceholder which is intended to be shown as additional visual hint when there is no selection for the dropdown.

## How Has This Been Tested?

Tested with Create React App TS project and Styleguidist using both mouse and keyboard navigation and VoiceOver for macOS. Tested using all supported browsers on Mac. Still needs testing with Win10 latest Edge and NVDA.

## Release notes
- Name prop is now called visual placeholder and is ignored by screen readers. Label has been added as mandatory prop.

### CODE REVIEW

File: src/core/Dropdown/Dropdown.test.tsx [hunk 1]
```diff
@@ -1,36 +1,132 @@
...
-      'aria-valid-attr-value': {
-        enabled: false,
+describe('Dropdown with additional aria-label', () => {
+  const additionalLabelProps: DropdownProps = {
+    ...dropdownProps,
+    'aria-labelledby': 'additional-label-id',
+  };
+  const DropdownWithExtraLabel = TestDropdown(additionalLabelProps);
+
+  it('should have aria-labelledby composed from labe-id and provided aria-labelledby prop', () => {
```
Comment (LJKaski) [inline src/core/Dropdown/Dropdown.test.tsx:96]: Small typo. labe-id --> label-id

File: src/components/Dropdown/Dropdown.tsx [hunk 2]
```diff
@@ -96,42 +125,79 @@ export class Dropdown extends Component<DropdownProps> {
...
-      className: classnames(buttonClassName, dropdownButtonProps.className),
+      id: buttonId,
+      className: classnames(
+        dropdownClassNames.button,
+        dropdownButtonProps.className,
+      ),
+      'aria-labelledby': `${
+        !!ariaLabelledBy ? `${ariaLabelledBy} ` : ''
+      }${labelId}`,
     };
```
Comment (LJKaski) [inline src/components/Dropdown/Dropdown.tsx:166]: I'd prefer to do this above with the id logic to keep the props clean and readable.

File: src/components/Dropdown/Dropdown.tsx [hunk 3]
```diff
@@ -42,13 +51,33 @@ type OptionalMenuPopoverProps = {
...
 
+type DropdownLabel = 'hidden' | 'top';
+
 export interface DropdownProps {
-  /** Name to show for the dropdown */
-  name: ReactNode;
-  /** Change name by selection
+  /** id */
+  id?: string;
+  /** Label */
```
[the remark this answers is not in the document]
Comment (aappoalander) [reply src/components/Dropdown/Dropdown.tsx:59]: For checkbox label is only children and the description is "Label element content" which is as ambiguous as the description for this label. Not sure how should we better describe label. Id is now copied from other comps using the uuid4 approach.

File: src/core/Dropdown/Dropdown.test.tsx (hunk 1 above)
Comment (aappoalander) [reply src/core/Dropdown/Dropdown.test.tsx]: Fixed.

File: src/components/Dropdown/Dropdown.tsx (hunk 2 above)
Comment (aappoalander) [reply src/components/Dropdown/Dropdown.tsx]: Fixed. Added a couple of empty lines to improve readability also.

File: src/components/Dropdown/Dropdown.tsx (hunk 3 above)
[the remark this answers is not in the document]
Comment (aappoalander) [reply src/components/Dropdown/Dropdown.tsx]: These should now be better. Added clarification for the Label as well to indicate that the label is for the whole component.

File: src/components/Dropdown/Dropdown.tsx [hunk 4]
```diff
@@ -42,13 +51,33 @@ type OptionalMenuPopoverProps = {
...
+  labelText: string;
+  /** Custom props for label container */
+  labelProps?: DropdownLabelProps;
+  /** Custom props for Label text element */
+  labelTextProps?: ParagraphProps;
+  /** Label displaymode -
+   * top: show above, hidden: use only for screenreader
+   * @default visible
```
[the remark this answers is not in the document]
Comment (aappoalander) [reply src/components/Dropdown/Dropdown.tsx]: Fixed.

File: src/components/Dropdown/Dropdown.tsx [hunk 5]
```diff
@@ -42,13 +51,33 @@ type OptionalMenuPopoverProps = {
...
+   */
+  labelMode?: DropdownLabel;
+  /**
+   * Additional label id. E.g. form group label.
+   * Used in addition to labelText for screen readers.
+   */
+  'aria-labelledby'?: string;
+  /** visual hint to show if nothing is selected */
```
[the remark this answers is not in the document]
Comment (aappoalander) [reply src/components/Dropdown/Dropdown.tsx]: Fixed.
